src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.Buffer.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigIntegerCalculator
+    {
+        // To spare memory allocations a buffer helps reusing memory!
+        // We just create the target array twice and switch between every
+        // operation. In order to not compute unnecessarily with all those
+        // leading zeros we take care of the current actual length.
+
+        // ATTENTION: always pass this by reference,
+        // it's a structure for performance reasons
+
+        internal struct BitsBuffer
+        {
+            private uint[] _bits;
+            private int _length;
+
+            public BitsBuffer(int length, uint value)
+            {
+                Debug.Assert(length >= 1);
+
+                _bits = new uint[length];
+                _length = value != 0 ? 1 : 0;
+
+                _bits[0] = value;
+            }
+
+            public BitsBuffer(int length, uint[] value)
+            {
+                Debug.Assert(value != null);
+                Debug.Assert(length >= ActualLength(value));
+
+                _bits = new uint[length];
+                _length = ActualLength(value);
+
+                Array.Copy(value, _bits, _length);
+            }
+
+            [SecuritySafeCritical]
+            public unsafe void MultiplySelf(ref BitsBuffer value,
+                                            ref BitsBuffer temp)
+            {
+                Debug.Assert(temp._length == 0);
+                Debug.Assert(_length + value._length <= temp._bits.Length);
+
+                fixed (uint* b = _bits, v = value._bits, t = temp._bits)
+                {
+                    if (_length < value._length)
+                    {
+                        Multiply(v, value._length,
+                                 b, _length,
+                                 t, _length + value._length);
+                    }
+                    else
+                    {
+                        Multiply(b, _length,
+                                 v, value._length,
+                                 t, _length + value._length);
+                    }
+                }
+
+                Apply(ref temp, _length + value._length);
+            }
+
+            [SecuritySafeCritical]
+            public unsafe void SquareSelf(ref BitsBuffer temp)
+            {
+                Debug.Assert(temp._length == 0);
+                Debug.Assert(_length + _length <= temp._bits.Length);
+
+                fixed (uint* b = _bits, t = temp._bits)
+                {
+                    Square(b, _length,
+                           t, _length + _length);
+                }
+
+                Apply(ref temp, _length + _length);
+            }
+
+            public void Reduce(ref FastReducer reducer)
+            {
+                _length = reducer.Reduce(_bits, _length);
+            }
+
+            [SecuritySafeCritical]
+            public unsafe void Reduce(uint[] modulus)
```

Comment:
Could you add some comments to explain what some of these functions are doing?  Some like MultipleSelf and SquareSelf are relatively self-explanatory, but this one is less clear, to me at least.